storage/storage_test.go

```diff
@@ -1344,3 +1360,149 @@ func TestWithEndpoint(t *testing.T) {
 	}
 	os.Setenv("STORAGE_EMULATOR_HOST", originalStorageEmulatorHost)
 }
+
+// Create a client using a combination of custom endpoint and
+// STORAGE_EMULATOR_HOST env variable and verify that raw.BasePath,
+// readHost and scheme are not changed by running operations such as upload
+func TestOperationsWithEndpoint(t *testing.T) {
+	originalStorageEmulatorHost := os.Getenv("STORAGE_EMULATOR_HOST")
+	defer os.Setenv("STORAGE_EMULATOR_HOST", originalStorageEmulatorHost)
+
+	addr, hClient, close := newTestServerAt(func(w http.ResponseWriter, r *http.Request) {
+		io.Copy(ioutil.Discard, r.Body)
+		fmt.Fprintf(w, "{}")
+	})
+	defer close()
+
+	testCases := []struct {
+		desc                string
+		CustomEndpoint      string
+		StorageEmulatorHost string
+	}{
+		{
+			desc:                "No endpoint or emulator host specified",
+			CustomEndpoint:      "",
+			StorageEmulatorHost: "",
+		},
+		{
+			desc:                "emulator host specified",
+			CustomEndpoint:      "",
+			StorageEmulatorHost: "https://" + addr,
+		},
+		{
+			desc:                "endpoint specified",
+			CustomEndpoint:      "https://" + addr,
+			StorageEmulatorHost: "",
+		},
+		{
+			desc:                "both emulator and endpoint specified",
+			CustomEndpoint:      "https://" + addr,
+			StorageEmulatorHost: "https://" + addr,
+		},
+	}
+	ctx := context.Background()
+	for _, tc := range testCases {
+		os.Setenv("STORAGE_EMULATOR_HOST", tc.StorageEmulatorHost)
+
+		c, err := NewClient(ctx, option.WithHTTPClient(hClient), option.WithEndpoint(tc.CustomEndpoint))
+		if err != nil {
+			t.Fatalf("error creating client: %v", err)
+		}
+		originalRawBasePath := c.raw.BasePath
+		originalReadHost := c.readHost
+		originalScheme := c.scheme
+
+		op := "bucket creation"
+
+		c.Bucket("test-bucket").Create(ctx, "pid", nil)
+		if err != nil {
+			t.Errorf("%s: %v", tc.desc, err)
+		}
+
+		if c.raw.BasePath != originalRawBasePath {
```

Comment:
I think you can just do this set of checks once at the end of the test case.

Reply:
Changed! My original idea was to provide information on which operation caused the change, but that's probably overkill and can easily be figured out if necessary by debugging.